File: examples/adt.py
from types import SimpleNamespace

Just = lambda x: lambda r: r.Just(x)
Nothing = lambda r: r.Nothing

def match(adt,**kwargs):
    return adt(SimpleNamespace(**kwargs))

def safeDivide(numerator, denominator):
    return Nothing if denominator == 0 else Just(numerator / denominator)

def displayResult(maybeValue):
    extractedValue = match(maybeValue,
      Just = lambda x: f'Result of division is {x}',
      Nothing = 'Handled divide by zero error'
    )
    print(extractedValue)

badDivision = safeDivide(5,0);
goodDivision = safeDivide(22, 7);

displayResult(badDivision);
displayResult(goodDivision);

File: examples/adt.pl
use strict;
use warnings;

sub Just {
    my ($x) = @_;
    sub {
        my ($handler) = @_;
        $handler->{Just}->($x);
    };
}

sub Nothing {
    sub {
        my ($handler) = @_;
        $handler->{Nothing};
    }
}

sub match {
    my ($value, $handler) = @_;
    $value->($handler);
}

sub safeDivide {
    my ($x, $y) = @_;
    $y == 0 ? Nothing() : Just($x/$y);
}

sub displayResult {
    my ($maybeValue) = @_;
    my $extractedValue = match($maybeValue, {
        Just => sub { my ($result) = @_; "Result of division is $result" },
        Nothing => "Handled divide by zero error"
    });
    print "$extractedValue\n";
}

my $badDivision = safeDivide(5,0);
my $goodDivision = safeDivide(22,7);

displayResult($badDivision);
displayResult($goodDivision);
